pkg/tfbridge/info_defaults.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tfbridge
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"unicode"
+
+	shim "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+
+	multierror "github.com/hashicorp/go-multierror"
+)
+
+// SetDefaults populates mappings such as the Resources mapping by computing sensible defaults in code. This is
+// especially useful for larger providers that have thousands of entries in Resources. Manually specified mappings
+// always take precedence: SetDefaults will never override an entry in the Resources map, only add new entries.
+func (info *ProviderInfo) SetDefaults(defaults ProviderInfoDefaults) error {
+	var errs multierror.Error
+	if defaults.Resources != nil {
+		if info.Resources == nil {
+			info.Resources = map[string]*ResourceInfo{}
+		}
+
+		if err := setDefaults(info.P.ResourcesMap(), info.Resources, defaults.Resources); err != nil {
+			errs.Errors = append(errs.Errors, fmt.Errorf("resources:\n%w", err))
+		}
+	}
+	if defaults.DataSources != nil {
+		if info.DataSources == nil {
+			info.DataSources = map[string]*DataSourceInfo{}
+		}
+		if err := setDefaults(info.P.DataSourcesMap(), info.DataSources, defaults.DataSources); err != nil {
+			errs.Errors = append(errs.Errors, fmt.Errorf("datasources:\n%w", err))
+		}
+	}
+	return errs.ErrorOrNil()
+}
+
+type ProviderInfoDefaults struct {
+	Resources   func(resourceToken string) (*ResourceInfo, error)
+	DataSources func(dataSourceToken string) (*DataSourceInfo, error)
+}
+
+func (d ProviderInfoDefaults) Unmappable(substring string) ProviderInfoDefaults {
```

Comment:
Just copying this from the previous design, curious did his turn out useful ? 

ALT is user explicitly doing:

```
Resources: func(resourceToken string) (*ResourceInfo, error) {
    if strings.Contans(resourceToken, "foo") {
        return nil, fmt.Errorf("Refusing to map...")
    }
    ...
}
```